api/stack_notification_batcher_test.go

```diff
@@ -0,0 +1,120 @@
+package api
+
+import (
+	"context"
+	"log/slog"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/buildkite/stacksapi"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+func TestNotificationBatcher_SendsBatchesPeriodically(t *testing.T) {
+	t.Parallel()
+
+	server := NewFakeAgentServer()
+	defer server.Close()
+
+	serverURL, _ := url.Parse(server.URL())
+	client, err := stacksapi.NewClient("fake-token", stacksapi.WithBaseURL(serverURL))
+	if err != nil {
+		t.Fatalf("NewClient() error = %v", err)
+	}
+
+	nb := newNotificationBatcher("test-stack", client, slog.Default())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if err := nb.start(ctx); err != nil {
+		t.Fatalf("start() error = %v", err)
+	}
+
+	// Add notifications
+	want := []stacksapi.StackNotification{
+		{JobUUID: "job-A", Detail: "detail A"},
+		{JobUUID: "job-B", Detail: "detail B"},
+		{JobUUID: "job-C", Detail: "detail C"},
+	}
+	for _, n := range want {
+		if err := nb.add(ctx, n); err != nil {
+			t.Fatalf("add() error = %v", err)
+		}
+	}
+
+	// Wait for ticker to flush (interval is 100ms)
+	time.Sleep(250 * time.Millisecond)
+
+	if got, wantLen := len(server.NotificationCalls), 1; got != wantLen {
+		t.Fatalf("number of notification calls = %d, want %d", got, wantLen)
+	}
+
+	if diff := cmp.Diff(want, server.NotificationCalls[0], cmpopts.IgnoreFields(stacksapi.StackNotification{}, "Timestamp")); diff != "" {
+		t.Errorf("notifications mismatch (-want +got):\n%s", diff)
+	}
+
+	cancel()
+	nb.waitDone()
+}
+
+func TestNotificationBatcher_ChunksBatchesBySize(t *testing.T) {
+	t.Parallel()
+
+	server := NewFakeAgentServer()
+	defer server.Close()
+
+	serverURL, _ := url.Parse(server.URL())
+	client, err := stacksapi.NewClient("fake-token", stacksapi.WithBaseURL(serverURL))
+	if err != nil {
+		t.Fatalf("NewClient() error = %v", err)
+	}
+
+	nb := newNotificationBatcher("test-stack", client, slog.Default())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if err := nb.start(ctx); err != nil {
+		t.Fatalf("start() error = %v", err)
+	}
+
+	// Add more than maxNotificationsPerBatch (1000) notifications
+	for i := range 2500 {
+		if err := nb.add(ctx, stacksapi.StackNotification{
+			JobUUID: "job-" + string(rune('A'+i%26)),
+			Detail:  "detail",
+		}); err != nil {
+			t.Fatalf("add() error = %v", err)
```

Comment:
If this fails, it would be good to know how far through the 2500 it got and what the notification object actually was:

```suggestion
		note := stacksapi.StackNotification{
			JobUUID: "job-" + string(rune('A'+i%26)),
			Detail:  "detail",
		}
		if err := nb.add(ctx, note); err != nil {
			t.Fatalf("i=%d add(%v) error = %v", i, note, err)
```